### DISCUSSION TITLE
[Outdated] What other non-trivial examples could we implement?

### DISCUSSION THREAD

**Discussion by Mikolaj** [Mikolaj/mostly-harmless#13] [Ideas]
In particular, it would be great to collect and implement any examples that require special cases and extensions of other, Haskell and non-Haskell libraries, but may or may not require extensions for us. Even if it turns out we can implement them on the spot, when added to the testsuite, they'd prevent reckless refactoring, e.g., for optimization or to cheaply accomplish other, orthogonal generalizations. 

Ideally, the examples would be large enough to indicate if complex nesting of the new stuff works fine and small enough that we can realistically test them on non-trivial data. MNIST is good size.

Please extend the list of candidates and fill in concrete problem statements, with links:
* any interesting optimization problems not via ML? [Edit: it seems there are a few here, but the book is from 2002 and I don't have it: https://link.springer.com/book/10.1007%2F978-1-4613-0075-5]
* classic astronomical or physical problems?
* any nice problem requiring a recurrent neural net (stateful functions rather than recursion, see https://blog.jle.im/entry/purely-functional-typed-models-3.html)?
* anything special in NLP?
* generative art? transformative art?

Edit: are there any neural net kinds (or anything else that uses gradients) that struggle with their use of matrices/tensors? E.g., the tensors are sparse, or sometimes sparse and sometimes dense, so no implementation is adequate? That may be where our free-form scalar wrangling approach shines. However, I'm afraid such approaches are labelled as "not parallelisable" and dropped and for a good reason, so it may be hard to find them.

Edit2: now that we do have vectors and matrices and that computing gradients with them is almost free (most of the time is spend computing the value of the function and applying gradients to parameters), our niche widens. I think horde-ad could shine wherever neural networks are complex enough (e.g., because they are dynamically generated) that storing traces and computing gradients with backpropagation becomes the limiting factor in their training. For simple networks containing huge tensors, I don't ever expect that. But can we find such cases with huge networks containing many medium size tensors? Ideally, if the networks change their shape continuously in an algorithmic way so that we don't have to code huge examples. but can let them generate themselves.

**Comment by Mikolaj** [comment]
Since we already have the scaffolding for MNIST data, here are a few papers that describe interesting neural networks that classify MNIST. I've chosen those that are somehow complex, but don't require pre-processing of data in order to facilitate an initial unsupervised pre-training phase or in order to make the training itself better. Sadly, all are tensor-based, probably because if a method works badly on tensors, it's not discussed nor sent to us, but trashed immediately.

* http://yann.lecun.com/exdb/publis/pdf/lecun-98.pdf (wait until we have arrays: this net requires MNIST data as a matrix, not a vector, and uses sub-matrix operations)
* https://www.microsoft.com/en-us/research/publication/best-practices-for-convolutional-neural-networks-applied-to-visual-document-analysis/?from=http%3A%2F%2Fresearch.microsoft.com%2Fapps%2Fpubs%2F%3Fid%3D68920 (similar to above, possibly not enough implementation details to replicate)
* https://github.com/Coderx7/SimpleNet (as above, but probably enough details)

Edit: after a closer look, these are all convolutional neural networks, which means they require data as a matrix (for fully connected networks that we implemented so far, it's enough to keep the data in a vector, which we do, not having array operations yet). It seems, the newer the paper, the more complex the network, but the more the paper advertises its simplicity.

**Reply by Mikolaj** [reply]
It seems not only convolutional, but also recurrent neural nets are getting great results in MNIST. And recurrent nns don't require matrices and work fine without pre-processing data, so that may be the best fit at this stage. I haven't chosen a paper or blog post to follow yet, so let me know if there's any particularly simple and well presented. Doesn't have to be new or have top MNIST scores.

**Reply by Mikolaj** [reply]
I've implemented a basic (no LSTM nor anything) RNN for MNIST and managed to train it (only after implementing Adam gradient descent *and* narrowly fine-tuning the variance of random initial parameter values). One of the implementations uses vectors solely, but a version using matrices is somewhat better and a version that stores mini-batches of data in matrices is the best of all (and requires fully general matrix multiplication and a lot of extra array operations). Next is CNN (https://github.com/Mikolaj/horde-ad/issues/9).

**Reply by Mikolaj** [reply]
For LSTM recurrent neural networks, once we have a naive implementation to verify low-level hacks against, bindings to https://github.com/libxsmm/libxsmm/tree/master/samples/deeplearning/lstmdriver may offer good performance (though it's worth making sure tensorflow or pythorch also uses it, so that we depend on something mainstream).

**Comment by Mikolaj** [comment]
A package on Hackage that uses ad and backprop to compute gradients: https://hackage.haskell.org/package/nonlinear-optimization (and another one, but it doesn't seem actively developed: https://github.com/kawu/sgd).

However, I don't think they are keen on performance or they'd use unboxed vectors (though then the library ad would not work for them). What we'd need is users keen on performance and with freely available use cases that we can benchmark with. I guess it's worth sweeping through Hackage for all reverse dependencies of ad and backprop (and mentions in readme and example code, because some package assume users would link to ad or backprop only in their code).

**Comment by Mikolaj** [comment]
@awf mentioned ResNet and networks with attention as possible next milestones after fully fledged convolution is implemented, in the deep learning track.

Edit: The main idea of ResNet, namely factoring in the output from a couple of previous layers, doesn't require any extensions to our machinery and there is no obvious extension that could speed it up. To implement ResNet18 for MNIST we'd need to extend our convolution to specific padding, strides, etc., and also implement batch normalization and AvgPool. Something to do after we get FFI to the convolution C or GPU libraries, so that we can test in reasonable time.

Edit2: See also https://github.com/Mikolaj/horde-ad/issues/50 and other tickets to which this discussion has been moved.
